test/Interop/Cxx/function/template-errors.swift

```diff
@@ -0,0 +1,39 @@
+// RUN: not %target-swift-emit-sil %s -I %S/Inputs -enable-cxx-interop 2>&1 | %FileCheck %s
+
+// README: If you just added support for protocol composition to the
+// ClangTypeConverter, please update this test to use a different type that we
+// don't support so the error messages here are still tested.
+
+
+import FunctionTemplates
+
+// Use protocol composition to create a type that we cannot (yet) turn into a clang::QualType.
+protocol A { }
+protocol B { }
+protocol C { }
+
+// CHECK: error: could not generate C++ types from the generic Swift types provided. The following Swift type(s) provided to 'passThrough' were unable to be converted: A & B.
+public func caller1(x: A & B) -> A & B {
+  return passThrough(x)
+}
+
+// CHECK: error: could not generate C++ types from the generic Swift types provided. The following Swift type(s) provided to 'addTwoTemplates' were unable to be converted: A & B, A & C.
+public func caller2(x: A & B, y: A & C) -> A & B {
+  return addTwoTemplates(x, y)
+}
+
+// Make sure we emit an error and don't crash when failing to instanciate a function.
+// CHECK: error: no matching function for call to 'takesString'
+// CHECK: note: in instantiation of function template specialization 'expectsString<long>' requested here
```

Comment:
I've discovered an unrelated issue. Currently, this is how we translate the following types (format: `swift` -> `cxx`): `Int` -> `long`, `Int32` -> `int`, and `Int64` -> `long long`. This is incorrect. (At least on the platform I'm working on) `Int` and `Int64` should translate to the same thing. Additionally, while not technically incorrect, I think if `Int32` translates to `int`, `Int64` should translate to `long` (not `long long`).

(semi-unrelated) Once this patch lands I can add some tests like this:
```
template<class T> size_t getSize(T) { return sizeof(T); }

func testSizesMatch<T>(_ x: T) { 
  expectEqual(MemoryLayout.size(ofValue: x), getSize(x));
}

testSizesMatch(0 as Int)
testSizesMatch(0 as Int32)
testSizesMatch(0 as Int64)
// ...
```

Which will be really nice and dynamic.

Reply:
> This is incorrect. (At least on the platform I'm working on)

What platform are you working on? Is there a loss of bits going from Swift -> C++ or the other way around?

> Additionally, while not technically incorrect, I think if Int32 translates to int, Int64 should translate to long (not long long).

`long` is not guaranteed to be at least 64-bits, so we can't do that. Windows uses 32-bit `long`.

https://en.cppreference.com/w/cpp/language/types

```
64 bit systems:
- LLP64 or 4/4/8 (int and long are 32-bit, pointer is 64-bit): Win64 API
- LP64 or 4/8/8 (int is 32-bit, long and pointer are 64-bit): Unix and Unix-like systems (Linux, macOS)
```

Reply:
> long is not guaranteed to be at least 64-bits, so we can't do that. Windows uses 32-bit long.

You're right. We can't do that. 

> What platform are you working on? Is there a loss of bits going from Swift -> C++ or the other way around?

I'm working on macOS. Turns out `long == long long` so we're all good. My bad, sorry for the false alarm. I got tripped up when it looked like one 64 bit integer was half the other, but it turns out they were both the same (correct) size.